homeassistant/components/hassio/__init__.py

```diff
@@ -684,16 +683,21 @@ async def _async_update_data(self) -> dict[str, Any]:
                 async_register_os_in_dev_reg(
                     self.entry_id, self.dev_reg, new_data["os"]
                 )
-            return new_data
 
         # Remove add-ons that are no longer installed from device registry
-        if removed_addons := list(set(self.data["addons"]) - set(new_data["addons"])):
-            async_remove_addons_from_dev_reg(self.dev_reg, removed_addons)
+        supervisor_addon_devices = {
+            list(device.identifiers)[0][1]
+            for device in self.dev_reg.devices.values()
+            if self.entry_id in device.config_entries
+            and device.model == SupervisorEntityModel.ADDON
+        }
+        if stale_addons := supervisor_addon_devices - set(new_data["addons"]):
+            async_remove_addons_from_dev_reg(self.dev_reg, stale_addons)
 
         # If there are new add-ons, we should reload the config entry so we can
         # create new devices and entities. We can return an empty dict because
         # coordinator will be recreated.
-        if list(set(new_data["addons"]) - set(self.data["addons"])):
+        if self.data and list(set(new_data["addons"]) - set(self.data["addons"])):
```

Comment:
Looks like we don't need to copy to list.

Reply:
Did not even notice that, will do a followup 👍

Reply:
<https://github.com/home-assistant/core/pull/52248>